extensions/amp-ad/0.1/amp-ad-batch-manager.js

```diff
@@ -0,0 +1,160 @@
+/**
+ * Copyright 2016 The AMP HTML Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS-IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+import {user} from '../../../src/log';
+import {xhrFor} from '../../../src/xhr';
+
+/** @const {number} The number of milliseconds between checks to see if the response has arrived */
+const AD_BATCH_TICK_LENGTH = 50;
+
+/** @const {number} The number of ticks to wait for a response from the ad server before timing out */
+const AD_BATCH_TIMEOUT = 100;
+
+/** @const {!string} Tag name for ad batch manager implementation. */
+export const TAG_AD_BATCH_MANAGER = 'amp-ad-batch-manager';
+
+/**
+  * Get a batch manager that will be used to batch together elements with the same data-url
+  * Note a side effect: the first ad with a given data-url to be processed will have its isBatchMaster_ variable set to true
+  * @param {Element} e The element which is to be batched
+  * @returns {AmpAdBatchManager}
+  */
+export function getBatchManager(e) {
+  // If this is our first imagead, create a map of responses for each value of 'data-url'
+  // This allows ads from multiple ad servers on the same page
+  if (!(e.win.hasOwnProperty('imageadBatchManagers'))) {
+    // Create an array of batch mangers. There will be one for each URL.
+    e.win.imageadBatchManagers = {};
+  }
+  if (!(e.url_ in e.win.imageadBatchManagers)) {
```

Comment:
please don't access private field (which potentially could be minified).

Instead, you might just pass data necessary data in the `getBachManager(data)`